Schastlivtsev.nikita

NikitaS4 opened the pull request.

NikitaS4 committed:
Add sources for lab1, script for compile and config file example

NikitaS4 committed:
Merge pull request #1 from NikitaS4/lab1

Add sources for lab1, script for compile and config file example

pv6 commented on schastlivtsev.nikita/lab1/DaemonCreator.cpp:
```diff
@@ -0,0 +1,82 @@
+void DaemonCreator::createDaemon(SettingsManager& settingsManager) {
+    forkWrapper(); // first fork
+
+    if (setsid() < 0) // new session
+        exit(1); // session creation failed
```
Почему exit(1), а не исключение?

NikitaS4 replied:
Исправил, теперь исключение

pv6 commented on schastlivtsev.nikita/lab1/DaemonCreator.cpp:
```diff
@@ -0,0 +1,82 @@
+
+    if (pid < 0) // error in fork
+        throw std::runtime_error("Fork has not been created (PID < 0)");
+    else if (pid > 0) { // kill parent
+        closelog();
+        exit(0);
```
Не стоит использовать exit(0) для остановки родительского процесса так как:
- Родительский процесс, в общем случае, может и продолжать работу после запуска демона
- exit сразу закрывает программу, без развертывания стека и вызова соответствующих деструкторов. Так что в родительском процессе происходит утечка памяти

Надо обернуть весь запуск демона (включая создание Логгера, парсинг конфига и т.д.) в одну функцию, вызываемую в main, а после неё родительский процесс уже либо завершается, либо в общем случае делает что ему дальше нужно.

NikitaS4 replied:
Не знал, что происходит утечка памяти, читал гайды для кода на Си. Исправил, теперь цепочкой `return`'ов исполнение доходит до `main`, где корректно завершается.

pv6 commented on schastlivtsev.nikita/lab1/Helpers.cpp:
```diff
@@ -0,0 +1,20 @@
+#include "Helpers.h"
+#include <cstring>
+
+void Helpers::strToCStr(char** dest, std::string& src) {
+    if (*dest) {
+        delete [] (*dest);
+        *dest = nullptr;
+    }
+    (*dest) = new char[src.length() + 1];
+    strcpy(*dest, src.c_str());
+}
+
+void Helpers::copyCStr(char** dest, char const * const src) {
+    if (*dest) {
+        delete [] (*dest);
+        *dest = nullptr;
+    }
+    (*dest) = new char[strlen(src) + 1];
+    strcpy(*dest, src);
+}
```
Почему не используете std::string для хранения строк?

NikitaS4 replied:
Хотелось использовать указатели в `Settings`, чтобы с помощью проверки на `nullptr` знать, когда соответствующее поле не заполнено.

pv6 commented on schastlivtsev.nikita/lab1/SignalHandler.h:
```diff
@@ -0,0 +1,22 @@
+        SignalHandler();
+        virtual ~SignalHandler();
+
+        void setSettingsManager(SettingsManager* settingsManager);
+        signalHandlerFunc getSigtermHandler();
+        signalHandlerFunc getSighupHandler();
```
Все эти функции могут быть static. В таком случае класс SignalHandler лучше сделать синглтоном.

NikitaS4 replied:
Сделал эти функции `static`. Нужды в том, чтобы это был синглтон, не оказалось, класс сделал полностью статическим.

pv6 commented on schastlivtsev.nikita/lab1/main.cpp:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        // create settings manager instance and pass config file to parse
+        SettingsManager settingsManager;
+        settingsManager.setConfig(config);
+
+        ProcedureBoss procBoss; // create procedure manager who launches procedure periodically
+        LogLogger logLogger; // create log logger to log logs (main daemon procedure executor)
+        settingsManager.linkEveryone(&procBoss, &logLogger); // link to the settings manager
+        settingsManager.updateSettings(); // parse config and set values to start
```
Все эти действия следовало бы делать только в дочернем процессе

NikitaS4 replied:
Поправил, поместил по максимуму всё это в дочерний процесс. Создание экземпляра `SettingsManager` не удалось избежать, так как он нужен для создания обработчиков сигналов. Однако его "честную" инициализацию (чтение config'а) отложил до дочернего процесса.

pv6 commented on schastlivtsev.nikita/lab1/SignalHandler.cpp:
```diff
@@ -0,0 +1,62 @@
+namespace {
+    static SettingsManager* s_settingsManager = nullptr;
+
+    static void sigTermHandler(int) {
+        syslog(LOG_INFO, "SIGTERM caught, exiting");
+        SignalHandler::removePidFile(s_settingsManager->getPidFile());
+        exit(0);
+    }
```
exit(0) здесь опять же приведет к потере памяти

NikitaS4 replied:
Спасибо, поправил.

NikitaS4 committed:
Remove exit(0) calls with main 'return'
Make SettingsManager class static (helper)
Moved config parsing to the stage where the last fork works only
Remove exit(1) with exception throwing

NikitaS4 committed:
Merge pull request #2 from NikitaS4/lab1

Correct mistakes
